govcd/adminvdc.go

```diff
@@ -23,6 +25,174 @@ func NewAdminVdc(cli *Client) *AdminVdc {
 	}
 }
 
+// vdcVersionedFuncs holds interface of functions
+type vdcVersionedFuncs struct {
+	SupportedVersion string
+	CreateVdc        func(adminOrg *AdminOrg, vdcConfiguration *types.VdcConfiguration) (*Vdc, error)
+	CreateVdcAsync   func(adminOrg *AdminOrg, vdcConfiguration *types.VdcConfiguration) (Task, error)
+	UpdateVdc        func(adminVdc *AdminVdc) (*AdminVdc, error)
+	UpdateVdcAsync   func(adminVdc *AdminVdc) (Task, error)
+}
+
+// VDC function mapping for API version 29.0 (from vCD 9.1)
+var vdcVersionedFuncsV90 = vdcVersionedFuncs{
+	SupportedVersion: "29.0",
+	CreateVdc:        createVdc,
+	CreateVdcAsync:   createVdcAsync,
+	UpdateVdc:        updateVdc,
+	UpdateVdcAsync:   updateVdcAsync,
+}
+
+// VDC function mapping for API version 32.0 (from vCD 9.7)
+var vdcVersionedFuncsV97 = vdcVersionedFuncs{
+	SupportedVersion: "32.0",
+	CreateVdc:        createVdcV97,
+	CreateVdcAsync:   createVdcAsyncV97,
+	UpdateVdc:        updateVdcV97,
+	UpdateVdcAsync:   updateVdcAsyncV97,
+}
+
+// VDC function mapping by vDC version
+var vdcVersionedFuncsByVcdVersion = map[string]vdcVersionedFuncs{
+	"vdc9.0":  vdcVersionedFuncsV90,
+	"vdc9.1":  vdcVersionedFuncsV90,
+	"vdc9.5":  vdcVersionedFuncsV90,
+	"vdc9.7":  vdcVersionedFuncsV97,
+	"vdc10.0": vdcVersionedFuncsV97,
+}
+
+// GetAdminVdcByName function uses a valid VDC name and returns a admin VDC object.
+// If no VDC is found, then it returns an empty VDC and no error.
+// Otherwise it returns an empty VDC and an error.
+// Deprecated: Use adminOrg.GetAdminVDCByName
+func (adminOrg *AdminOrg) GetAdminVdcByName(vdcname string) (AdminVdc, error) {
+	for _, vdcs := range adminOrg.AdminOrg.Vdcs.Vdcs {
+		if vdcs.Name == vdcname {
+			adminVdc := NewAdminVdc(adminOrg.client)
+			_, err := adminOrg.client.ExecuteRequest(vdcs.HREF, http.MethodGet,
+				"", "error getting vdc: %s", nil, adminVdc.AdminVdc)
+			return *adminVdc, err
+		}
+	}
+	return AdminVdc{}, nil
+}
+
+// GetAdminVDCByHref retrieves a VDC using a direct call with the HREF
+func (adminOrg *AdminOrg) GetAdminVDCByHref(vdcHref string) (*AdminVdc, error) {
+
+	adminVdc := NewAdminVdc(adminOrg.client)
+
+	// We are executing below request with a specific API version in the header, because we want to retrieve the most
+	// available fields in AdminVdc which vCD provides, but also which our code understands. As we can't blindly use
+	// the latest version, we're limiting the highest used version to the one we support with
+	// the GetSpecificApiVersionOnCondition(...) function. Specifically, the API version 32 returns
+	// two additional fields: IncludeMemoryOverhead and IsElastic for Flex allocation
+	_, err := adminOrg.client.ExecuteRequestWithApiVersion(vdcHref, http.MethodGet,
```

Comment:
If possible, I'd like this function to be implemented similarly to the others defined in `vdcVersionedFuncs`, i.e. one function per version.

Reply:
Possible, yet it won't bring any value just code repetition

Reply:
Code repetition is when we repeat the same code in various places. I am saying we should have a GET for 9.0 to 9.5 and another for 9.7+, same as what we did for Create and Update

Reply:
I don't see any value here until more changes needed. Now I will create same functions for sake of creating and have to make tests, and so on.